src/Microsoft.ML.Data/Training/TrainerInputBase.cs

```diff
@@ -0,0 +1,197 @@
+        /// </summary>
+        [BestFriend]
+        [Argument(ArgumentType.LastOccurenceWins, HelpText = "Whether learner should cache input training data", ShortName = "cache", SortOrder = 6, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        internal CachingOptions Caching = CachingOptions.Auto;
+    }
+
+    /// <summary>
+    /// The base class for all learner inputs that support a Label column.
+    /// </summary>
+    [TlcModule.EntryPointKind(typeof(CommonInputs.ITrainerInputWithLabel))]
+    public abstract class TrainerInputBaseWithLabel : TrainerInputBase
+    {
+        /// <summary>
+        /// Column to use for labels.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, HelpText = "Column to use for labels", ShortName = "lab", SortOrder = 3, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string LabelColumn = DefaultColumnNames.Label;
+    }
+
+    // REVIEW: This is a known antipattern, but the solution involves the decorator pattern which can't be used in this case.
+    /// <summary>
+    /// The base class for all learner inputs that support a weight column.
+    /// </summary>
+    [TlcModule.EntryPointKind(typeof(CommonInputs.ITrainerInputWithWeight))]
+    public abstract class TrainerInputBaseWithWeight : TrainerInputBaseWithLabel
+    {
+        /// <summary>
+        /// Column to use for example weight.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, HelpText = "Column to use for example weight", ShortName = "weight", SortOrder = 4, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string WeightColumn = null;
+    }
+
+    /// <summary>
+    /// The base class for all unsupervised learner inputs that support a weight column.
+    /// </summary>
+    [TlcModule.EntryPointKind(typeof(CommonInputs.IUnsupervisedTrainerWithWeight))]
+    public abstract class UnsupervisedTrainerInputBaseWithWeight : TrainerInputBase
+    {
+        /// <summary>
+        /// Column to use for example weight.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, HelpText = "Column to use for example weight", ShortName = "weight", SortOrder = 4, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string WeightColumn = null;
+    }
+
+    [TlcModule.EntryPointKind(typeof(CommonInputs.ITrainerInputWithGroupId))]
+    public abstract class TrainerInputBaseWithGroupId : TrainerInputBaseWithWeight
+    {
+        /// <summary>
+        /// Column to use for example groupId.
+        /// </summary>
+        [Argument(ArgumentType.AtMostOnce, Name = "GroupIdColumn", HelpText = "Column to use for example groupId", ShortName = "groupId", SortOrder = 5, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly)]
+        public string GroupIdColumn = null;
+    }
+
+    [BestFriend]
+    internal static class TrainerEntryPointsUtils
```

Comment:
>TrainerEntryPointsUtils [](start = 26, length = 23)

As near as I can tell this is specific to entry-point applications and infrastructure, and should remain with the internal classes inside entry-points.